include/nuttx/can/sja1000.h

```diff
@@ -0,0 +1,163 @@
+/****************************************************************************
+ * include/nuttx/can/sja1000.h
+ *
+ * SJA1000 CAN driver based on esp32c3_twai.h
+ *
+ * License header retained from original source.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+#ifndef __DRIVERS_INCLUDE_CAN_SJA1000_H
+#define __DRIVERS_INCLUDE_CAN_SJA1000_H
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Public Types
+ ****************************************************************************/
+
+/* CAN hardware-dependent bit-timing constant
+ *
+ * Used for calculating and checking bit-timing parameters
+ */
+
+struct sja1000_can_bittiming_const
```

Comment:
```suggestion
struct sja1000_can_bittiming_s
```

Reply:
I would be most pleased, if the original `can_bittiming_const` or `can_bittiming_const_s` can be kept. This structure  represents constant/fixed part of timing data given by CAN controller registers and internal logic implementation. It has been designed such way to allow parametrize all existing controllers (around 2005) and has been proved to do so quite reasonable. See use in the Linux kernel
https://elixir.bootlin.com/linux/latest/A/ident/can_bittiming_const
Basically copying numbers from the Linux kernel would allow to resolve bittiming rules for quite broad range of controllers. As for the structure/its fields copyright, original version predates SocketCAN, see `can_calc_const_t ` in master's thesis https://wiki.control.fel.cvut.cz/mediawiki/images/3/37/Dp_2006_marek_stanislav.pdf which I have supervised and proposed algorithms for calculation based on my LinCAN work.

There should be companion structure like Linux kernel `can_bittiming_const_s` https://elixir.bootlin.com/linux/latest/source/include/uapi/linux/can/netlink.h#L31 which keeps data calculated/specific for given bitrate. It is newer than my initial proposal, because in that has been computed parameters returned individually
```
int can_auto_baud(int *pbrp, int *ptseg1, int *ptseg2, long baud_rate,
                  int sampl_pt, long clock, can_calc_const_t * c) ```